streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala

```diff
@@ -190,283 +281,150 @@ class WriteAheadLogSuite extends SparkFunSuite with BeforeAndAfter {
-    data.foreach { item =>
-      manualClock.advance(clockAdvanceTime)
-      wal.write(item, manualClock.getTimeMillis())
-    }
-    if (closeLog) wal.close()
-    wal
-  }
-
-  /** Read data from a segments of a log file directly and return the list of byte buffers. */
-  def readDataManually(segments: Seq[FileBasedWriteAheadLogSegment]): Seq[String] = {
-    segments.map { segment =>
-      val reader = HdfsUtils.getInputStream(segment.path, hadoopConf)
-      try {
-        reader.seek(segment.offset)
-        val bytes = new Array[Byte](segment.length)
-        reader.readInt()
-        reader.readFully(bytes)
-        val data = Utils.deserialize[String](bytes)
-        reader.close()
-        data
-      } finally {
-        reader.close()
+  private def writeBlockingPromise(wal: WriteAheadLog): Promise[Any] = {
+    // we would like to block the write so that we can queue requests
+    val promise = Promise[Any]()
+    when(wal.write(any[ByteBuffer], any[Long])).thenAnswer(
+      new Answer[FileBasedWriteAheadLogSegment] {
+        override def answer(invocation: InvocationOnMock): FileBasedWriteAheadLogSegment = {
+          Await.ready(promise.future, 4.seconds)
+          walHandle
+        }
       }
-    }
+    )
+    promise
   }
 
-  /** Read all the data from a log file directly and return the list of byte buffers. */
-  def readDataManually(file: String): Seq[String] = {
-    val reader = HdfsUtils.getInputStream(file, hadoopConf)
-    val buffer = new ArrayBuffer[String]
-    try {
-      while (true) {
-        // Read till EOF is thrown
-        val length = reader.readInt()
-        val bytes = new Array[Byte](length)
-        reader.read(bytes)
-        buffer += Utils.deserialize[String](bytes)
-      }
-    } catch {
-      case ex: EOFException =>
-    } finally {
-      reader.close()
+  test("BatchedWriteAheadLog - name log with aggregated entries with the timestamp of last entry") {
+    val wal = new BatchedWriteAheadLog(fileBasedWAL, sparkConf)
+    // block the write so that we can batch some records
+    val promise = writeBlockingPromise(fileBasedWAL)
+
+    val event1 = "hello"
+    val event2 = "world"
+    val event3 = "this"
+    val event4 = "is"
+    val event5 = "doge"
+
+    eventFuture(wal, event1, 3L) // 3 will automatically be flushed for the first write
+    // rest of the records will be batched while it takes 3 to get written
+    eventFuture(wal, event2, 5L)
+    eventFuture(wal, event3, 8L)
+    eventFuture(wal, event4, 12L)
+    eventFuture(wal, event5, 10L)
+    promise.success(true)
+
+    eventually(Eventually.timeout(1 second)) {
+      verify(fileBasedWAL, times(1)).write(any[ByteBuffer], meq(3L))
+      // the file name should be the timestamp of the last record, as events should be naturally
+      // in order of timestamp, and we need the last element.
+      verify(fileBasedWAL, times(1)).write(any[ByteBuffer], meq(10L))
```

Comment:
Shouldnt you verify that the futures have been completed with some expected WAL segment?